plugins/georeplication/rest.go

```diff
@@ -0,0 +1,193 @@
+		restutils.SendHTTPError(w, http.StatusBadRequest, "Atleast one Slave host is required")
+		return
+	}
+
+	if req.SlaveVol == "" {
+		restutils.SendHTTPError(w, http.StatusBadRequest, "Slave volume name is required field")
+		return
+	}
+
+	// Check if Master volume exists and Matches with passed Volume ID
+	vol, e := volume.GetVolume(req.MasterVol)
+	if e != nil {
+		restutils.SendHTTPError(w, http.StatusNotFound, errors.ErrVolNotFound.Error())
+		return
+	}
+
+	// Check if Master Volume ID from store matches the input Master Volume ID
+	if !uuid.Equal(vol.ID, masterid) {
+		restutils.SendHTTPError(w, http.StatusBadRequest, "Master volume ID doesn't match")
+		return
+	}
+
+	// Fetch existing session details from Store, if same
+	// session exists then return error
+	geoSession, err := getSession(masterid.String(), slaveid.String())
+	if err == nil {
+		if !updateRequest {
+			restutils.SendHTTPError(w, http.StatusConflict, "Session already exists")
+			return
+		}
+	} else {
+		// Continue only if NotFound error, return if other errors like
+		// error while fetching from store or JSON marshal errors
+		if _, ok := err.(*ErrGeorepSessionNotFound); !ok {
+			restutils.SendHTTPError(w, http.StatusInternalServerError, err.Error())
+			return
+		}
+		// Fail if update request received for non existing Geo-rep session
+		if updateRequest {
+			restutils.SendHTTPError(w, http.StatusBadRequest, "geo-replication session not found")
+			return
+		}
+
+		geoSession = new(georepapi.GeorepSession)
+		geoSession.MasterID = masterid
+		geoSession.SlaveID = slaveid
+	}
+
+	// If Update request then it is not allowed to change master volume,
+	// slave volname, Master Volume name and ID is already verified earlier
+	// only validate for slave vol name here
+	if updateRequest && geoSession.SlaveVol != req.SlaveVol {
+		restutils.SendHTTPError(w, http.StatusBadRequest, "Slave Volume name can't be modified")
+		return
+	}
+
+	// Set/Update the input details
+	geoSession.MasterVol = req.MasterVol
+	geoSession.SlaveVol = req.SlaveVol
+	geoSession.SlaveHosts = req.SlaveHosts
+	geoSession.SlaveUser = req.SlaveUser
+
+	// Set Slave user as root if request doesn't contains user name
+	if geoSession.SlaveUser == "" {
+		geoSession.SlaveUser = "root"
+	}
+
+	geoSession.Workers = []georepapi.GeorepWorker{}
+	geoSession.Options = make(map[string]string)
+
+	// Transaction which updates the Geo-rep session
+	txn := transaction.NewTxn(reqID)
+	defer txn.Cleanup()
+
+	// Lock on Master Volume name
+	lock, unlock, err := transaction.CreateLockSteps(geoSession.MasterVol)
+	if err != nil {
+		restutils.SendHTTPError(w, http.StatusInternalServerError, err.Error())
+		return
+	}
+
+	// TODO: Transaction step function for setting Volume Options
+	// As a workaround, Set volume options before enabling Geo-rep session
+
+	txn.Nodes = vol.Nodes()
+	txn.Steps = []*transaction.Step{
+		lock,
+		{
+			DoFunc: "georeplication-create.Commit",
+			Nodes:  []uuid.UUID{gdctx.MyUUID},
```

Comment:
As this run only one node, is this made a transaction only for the locking ?

Reply:
There are some todos for Geo-rep create steps. I think I need to create a working directory in all Volume nodes before starting. I will work on those steps later.